packages/core/database/lib/entity-manager/relations-orderer.js

```diff
@@ -1,6 +1,107 @@
+  let needsSorting = false;
+  // Map to validate if relation is already in sortedConnect or DB.
+  const relationInInitialArray = initialArr.reduce((acc, rel) => ({ ...acc, [rel.id]: true }), {});
+  // Map to store the first index where a relation id is connected
+  const mappedRelations = connectArr.reduce((mapper, relation) => {
+    const adjacentRelId = relation.position?.before || relation.position?.after;
+
+    if (!relationInInitialArray[adjacentRelId] && !mapper[adjacentRelId]) {
+      needsSorting = true;
+    }
+
+    // If the relation is already in the array, throw an error
+    if (mapper[relation.id]) {
+      throw new InvalidRelationError(
+        `The relation with id ${relation.id} is already connected. ` +
+          'You cannot connect the same relation twice.'
+      );
+    }
+
+    return {
+      [relation.id]: { ...relation, computed: false },
+      ...mapper,
+    };
+  }, {});
+
+  // If we don't need to sort the connect array, we can return it as is
+  if (!needsSorting) return connectArr;
+
+  // Recursively compute in which order the relation should be connected
+  const computeRelation = (relation, relationsSeenInBranch) => {
+    const adjacentRelId = relation.position?.before || relation.position?.after;
+    const adjacentRelation = mappedRelations[adjacentRelId];
+
+    // If the relation has already been seen in the current branch,
+    // it means there is a circular reference
+    if (relationsSeenInBranch[adjacentRelId]) {
+      throw new InvalidRelationError(
+        'A circular reference was found in the connect array. ' +
+          'One relation is trying to connect before/after another one that is trying to connect before/after it'
+      );
+    }
+
+    // This relation has already been computed
+    if (mappedRelations[relation.id]?.computed) return;
+
+    // Relation does not have a before or after attribute or is in the initial array
+    if (!adjacentRelId || relationInInitialArray[adjacentRelId]) {
+      mappedRelations[relation.id].computed = true;
+      sortedConnect.push(relation);
+      return;
+    }
+
+    // Look if id is referenced elsewhere in the array
+    if (mappedRelations[adjacentRelId]) {
+      mappedRelations[relation.id].computed = true;
+      computeRelation(adjacentRelation, { ...relationsSeenInBranch, [relation.id]: true });
+      sortedConnect.push(relation);
+    } else if (strictSort) {
+      // If we reach this point, it means that the adjacent relation is not in the connect array
+      // and it is not in the database.
+      throw new InvalidRelationError(
+        `There was a problem connecting relation with id ${
+          relation.id
+        } at position ${JSON.stringify(
+          relation.position
+        )}. The relation with id ${adjacentRelId} needs to be connected first.`
+      );
+    } else {
+      // We are in non-strict mode so we can push the relation.
+      mappedRelations[relation.id].computed = true;
```

Comment:
As `mappedRelations[relation.id].computed` is always set to true, maybe we can do this:
```suggestion
      mappedRelations[relation.id].computed = true;

    // Relation does not have a before or after attribute or is in the initial array
    if (!adjacentRelId || relationInInitialArray[adjacentRelId]) {
      sortedConnect.push(relation);
      return;
    }

    // Look if id is referenced elsewhere in the array
    if (mappedRelations[adjacentRelId]) {
      computeRelation(adjacentRelation, { ...relationsSeenInBranch, [relation.id]: true });
      sortedConnect.push(relation);
    } else if (strictSort) {
      // If we reach this point, it means that the adjacent relation is not in the connect array
      // and it is not in the database.
      throw new InvalidRelationError(
        `There was a problem connecting relation with id ${
          relation.id
        } at position ${JSON.stringify(
          relation.position
        )}. The relation with id ${adjacentRelId} needs to be connected first.`
      );
    } else {
      // We are in non-strict mode so we can push the relation.
```

Reply:
💡oh yes, good one.